Add migration doc draft. (#91)
temporalio/temporal-worker-controller: docs/migration-guide.md

# Migrating to Temporal Worker Controller

> This guide helps teams migrate from their existing versioned worker deployment systems to the Temporal Worker Controller. It assumes you are already running versioned workers using Temporal's Worker Versioning feature and want to automate the management of these deployments.

Author: Yep, true, re-written from unversioned.

## Important Terminology

To avoid confusion, this guide uses specific terminology:

> - **Temporal Worker Deployment**: A logical grouping in Temporal (e.g., "payment-processor", "notification-sender")

Reviewer: nit: can we add the following definitions from the temporal docs so that we stay precise with the terms used here?

referring to:

> A Worker Deployment is a deployment or service across multiple versions. In a rainbow deployment, more than two active Deployment Versions can run at once.
A Worker Deployment Version is a version of a deployment or service. It can have multiple Workers polling on multiple Task Queues, but they all run the same build.
A Build ID, in combination with a Worker Deployment name, identifies a single Worker Deployment Version. 

feel free to disagree, and maybe after this information goes in the outer README this might be redundant we could also just point to that area if we don't wanna repeat ourselves

wdyt

Reviewer: btw took the defs out from - https://docs.temporal.io/production-deployment/worker-deployments/worker-versioning

    strategy: Manual  # Use Manual during migration, then switch to Progressive
  sunset:
    scaledownDelay: 1h
    deleteDelay: 24h
  template:
    spec:
      containers:
      - name: worker
        image: my-worker:v1.2.4  # Update this image to import/deploy versions
```

**Key Differences:**
- ✅ **Single CRD resource** manages multiple versions instead of multiple manual Kubernetes Deployments
- ✅ **Controller creates Kubernetes Deployments** automatically (one per version)
- ✅ **Automated lifecycle** handles registration, routing, and cleanup
- ✅ **Version history** tracked in the CRD resource status
- ✅ **Rollout strategies** automate traffic shifting between versions

## Migration Strategy

### The Single Resource Requirement

⚠️ **CRITICAL**: For each **Temporal Worker Deployment**, you MUST create exactly **ONE** `TemporalWorkerDeployment` CRD resource and import all existing versions into it. This is not optional - it's how the system works.

**Terminology Clarification:**
- **Temporal Worker Deployment**: A logical grouping in Temporal (e.g., "payment-processor")
- **`TemporalWorkerDeployment` CRD**: One Kubernetes custom resource per Temporal Worker Deployment

> - **Kubernetes `Deployment`**: Multiple k8s Deployments (one per version) managed by the controller

Reviewer: doesn't seem like we need this given we have a nice section at the top explaining these things to us

**Key Principles:**
- **One CRD resource per Temporal Worker Deployment** - Don't create multiple `TemporalWorkerDeployment` resources for the same logical worker

> - **Use Manual strategy during import** - Prevents unwanted automatic promotions

Reviewer: just wanted to point out that if we are referring to the different strategies over here, we should probably have them defined in the outer README which, as it stands, is going to be a giant encyclopedia of terms for a new reader haha

- **Import versions sequentially** - Update the same CRD resource to import each existing version
- **Enable automation last** - Switch to Progressive strategy only after migration is complete

### Recommended Migration Steps

1. **Install the controller** in your cluster
2. **Choose a non-critical worker** to migrate first
3. **Create a single TemporalWorkerDeployment** with Manual strategy
4. **Import existing versions one by one** by updating the image spec
5. **Clean up manual deployments** after confirming controller management
6. **Enable automated rollouts** by switching to Progressive strategy

> 7. **Repeat for remaining workers** one deployment at a time

Reviewer: nit: this seems to be the crux of this section and maybe we should leave lines between each point over here; feels to cluttered to eye rn wdyt

```yaml
# Version 1 (older, still has running workflows)
apiVersion: apps/v1
kind: Deployment
metadata:
  name: payment-processor-v1
spec:
  replicas: 2
  template:
    spec:
      containers:
      - name: worker
        image: payment-processor:v1.5.1
        env:
        - name: WORKER_BUILD_ID
          value: "v1.5.1"
---
# Version 2 (current production version)
apiVersion: apps/v1
kind: Deployment
metadata:
  name: payment-processor-v2
spec:
  replicas: 5
  template:
    spec:
      containers:
      - name: worker
        image: payment-processor:v1.5.2
        env:
        - name: WORKER_BUILD_ID
          value: "v1.5.2"
```

**Single `TemporalWorkerDeployment` CRD Resource (CRITICAL: Use Manual strategy):**
```yaml
apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: payment-processor
  labels:
    app: payment-processor
spec:
  replicas: 5
  workerOptions:
    connection: production-temporal
    temporalNamespace: production
  # IMPORTANT: Use Manual during migration to prevent unwanted promotions
  cutover:
    strategy: Manual
  sunset:
    scaledownDelay: 30m
    deleteDelay: 2h
  template:
    metadata:
      annotations:

>         prometheus.io/scrape: "true"

Reviewer: maybe we don't need to add this here since it is not super relevant to the migration side of things and seems prom heavy

⚠️ **CRITICAL**: Use a single `TemporalWorkerDeployment` CRD resource with `strategy: Manual` to import all existing versions.

1. **Create the `TemporalWorkerDeployment` CRD with the oldest version:**
   ```bash
   kubectl apply -f payment-processor-migration.yaml
   ```

2. **Wait for the first version to be registered:**
   ```bash
   # Monitor until status shows the version is registered and current
   kubectl get temporalworkerdeployment payment-processor -w
   ```
   
   The controller will create a Kubernetes `Deployment` resource (e.g., `payment-processor-v1.5.1`) for this version.

3. **Import the next version by updating the CRD resource:**
   ```bash
   # Edit the TemporalWorkerDeployment CRD to use the next image version
   kubectl patch temporalworkerdeployment payment-processor --type='merge' -p='{"spec":{"template":{"spec":{"containers":[{"name":"worker","image":"payment-processor:v1.5.2"}]}}}}'
   ```

4. **Wait for the new version to be registered:**
   ```bash
   # Monitor until the new version appears in status.targetVersion
   kubectl get temporalworkerdeployment payment-processor -o yaml
   ```

   You should see:
   ```yaml
   status:
     currentVersion:

>        versionID: "payment-processor.v1.5.1"

Reviewer: correct me if I am wrong, but this version will only be set as current if it is indeed current for the temporal server, right?

are we assuming it's the current version?

          memory: "1Gi"
          cpu: "500m"
        limits:
          memory: "2Gi" 
          cpu: "1"
```

**Pod Annotations and Labels:**
```yaml
template:
  metadata:
    annotations:
      prometheus.io/scrape: "true"
      prometheus.io/port: "9090"
    labels:
      team: payments
      environment: production
```

## Testing and Validation

### Pre-Migration Testing

1. **Test in non-production environment:**
   ```bash
   # Create test TemporalWorkerDeployment
   kubectl apply -f test-worker.yaml
   
   # Verify worker registration
   kubectl logs -l temporal.io/worker-deployment=test-worker
   ```

2. **Validate environment variables:**
   ```bash
   kubectl exec -it deployment/test-worker-v1 -- env | grep TEMPORAL
   ```

### Post-Migration Validation

1. **Check version status:**
   ```bash
   kubectl get temporalworkerdeployment -o wide
   ```

2. **Monitor version transitions:**
   ```bash
   kubectl get events --field-selector involvedObject.kind=TemporalWorkerDeployment
   ```

3. **Validate workflow routing:**
   - Start test workflows
   - Verify they're routed to correct versions
   - Check Temporal UI for version distribution

## Common Migration Patterns

> ⚠️ **Remember**: Each pattern below represents **separate Temporal Worker Deployments**. Each gets exactly **one** `TemporalWorkerDeployment` CRD resource.

Reviewer: haha this is so cursor-oriented (I also remember you mentioning you did while holding Cursor's hand); do we really need these cautions, feel like while doing a single pass I have come across them multiple times now

> ### Pattern 1: Microservices with Multiple Workers

Reviewer: I really like this section since it helps the reader go through a mental process of "is this for me"

metadata:
  name: notification-sender
spec:
  workerOptions:
    connection: production-temporal
    temporalNamespace: notifications
  cutover:
    strategy: Manual  # Use Manual during migration
  # ... rest of config
```

### Pattern 2: Environment-Specific Deployments

```yaml
# Production (use Manual during migration, then switch to Progressive)
apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: my-worker
  namespace: production
spec:
  workerOptions:
    connection: production-temporal
    temporalNamespace: production
  cutover:
    strategy: Manual  # Use Manual during migration, then switch to Progressive
---
# Staging (use Manual during migration, then switch to AllAtOnce)
apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: my-worker
  namespace: staging
spec:
  workerOptions:
    connection: staging-temporal
    temporalNamespace: staging
  cutover:
    strategy: Manual  # Use Manual during migration, then switch to AllAtOnce
```

## Troubleshooting

### Common Issues

**1. Workers Not Registering with Temporal**

*Symptoms:*
```
status:
  targetVersion:
    status: NotRegistered
```

*Solutions:*

> - Check worker logs for connection errors

Reviewer: nit: connection/initialization errors

- Verify TemporalConnection configuration
- Ensure TLS secrets are properly configured
- Verify network connectivity to Temporal server

**2. Version Stuck in Ramping State**

*Symptoms:*
```
status:
  targetVersion:
    status: Ramping
    rampPercentage: 5
```

*Solutions:*
- Check if gate workflow is configured and completing successfully
- Verify progressive rollout steps are reasonable

> - Check controller logs for errors

Reviewer: I know there is a section at the bottom about how one can look at the controller logs; we should place a to the section there

kubectl describe temporalworkerdeployment my-worker

# Check managed deployments
kubectl get deployments -l temporal.io/managed-by=temporal-worker-controller

# Check worker logs
kubectl logs -l temporal.io/worker-deployment=my-worker

# Check controller events
kubectl get events --field-selector involvedObject.kind=TemporalWorkerDeployment
```

### Getting Help

1. **Check controller logs** for error messages
2. **Review TemporalWorkerDeployment status** for detailed state information
3. **Verify Temporal server connectivity** from worker pods
4. **File issues** at the project repository with logs and configuration

## Migration Summary

🎯 **Remember the core principle**: 
- **One `TemporalWorkerDeployment` CRD per Temporal Worker Deployment**
- **Manual strategy during migration** to prevent unwanted promotions
- **Import existing versions sequentially** by updating the same CRD resource
- **Enable automation only after migration is complete**

**Resource Relationship:**
- **Before**: Multiple manual Kubernetes `Deployment` resources per worker
- **After**: One `TemporalWorkerDeployment` CRD → Controller creates multiple Kubernetes `Deployment` resources (one per version)

> This approach ensures a smooth transition from manual version management to controller automation without disrupting running workflows.

Reviewer: I think we can remove this section on the basis of redundancy

## Next Steps

After successful migration:

1. **Set up monitoring** for your TemporalWorkerDeployment resources
2. **Update CI/CD pipelines** to patch TemporalWorkerDeployment image specs instead of managing Deployments directly
3. **Configure alerting** on version transition failures
4. **Train your team** on the new deployment process (single resource updates vs multiple Deployment management)

> 5. **Document your specific configuration** patterns for future reference

Reviewer: this also does not feel technical focus and might just be team dependant; not sure if we wanna keep this

3. [Understanding the Differences](#understanding-the-differences)
4. [Migration Strategy](#migration-strategy)
5. [Step-by-Step Migration](#step-by-step-migration)
6. [Configuration Reference](#configuration-reference)
7. [Testing and Validation](#testing-and-validation)
8. [Common Migration Patterns](#common-migration-patterns)
9. [Troubleshooting](#troubleshooting)

## Why Migrate to Versioned Workflows

If you're currently running unversioned Temporal workflows, you may be experiencing challenges with deployments:

- **Deployment Risk**: Code changes can break running workflows if they're not backward compatible
- **Rollback Complexity**: Rolling back deployments can disrupt in-flight workflows
- **Workflow Determinism Issues**: Changes to workflow logic can cause non-deterministic errors

Versioned workflows with the Temporal Worker Controller solve these problems by:

- ✅ **Safe Deployments**: New versions run alongside old ones, ensuring running workflows complete successfully
- ✅ **Automated Rollouts**: Progressive rollout strategies reduce risk of new deployments
- ✅ **Easy Rollbacks**: Can instantly route new workflows back to previous versions

> - ✅ **Workflow Continuity**: Running workflows can continue on their original version until completion

Reviewer 2: would it make sense to link out to https://docs.temporal.io/production-deployment/worker-deployments/worker-versioning for the benefits (and if the docs are missing some of the things mentioned here, move that to the docs) instead of rehashing it here?

Reviewer 2: The first section of that doc gives a good overview of the benefits, and I feel like we don't need to repeat ourselves

## Prerequisites

Before starting the migration, ensure you have:

- ✅ **Unversioned Temporal workers**: Currently running workers without Worker Versioning
- ✅ **Kubernetes cluster**: Running Kubernetes 1.19+ with CustomResourceDefinition support
- ✅ **Basic worker configuration**: Workers connect to Temporal with namespace and task queue configuration
- ✅ **Administrative access**: Ability to install Custom Resource Definitions and controllers
- ✅ **Deployment pipeline**: Existing CI/CD system that can be updated to use new deployment method

### Current Environment Variables

Your workers are likely configured with basic environment variables like:

```bash

> TEMPORAL_HOST_PORT=your-temporal-namespace.tmprl.cloud:7233

Reviewer 2: ```suggestion
TEMPORAL_ADDRESS=your-temporal-namespace.tmprl.cloud:7233
```
changed recently to match SDK expectations!

TEMPORAL_NAMESPACE=your-temporal-namespace
# No TEMPORAL_DEPLOYMENT_NAME or TEMPORAL_WORKER_BUILD_ID yet
```

The controller will automatically add the versioning-related environment variables during migration.

## Understanding the Differences

### Before: Unversioned Workers

```yaml
# Single deployment, all workers run the same code
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-worker
spec:
  replicas: 3
  template:
    spec:
      containers:
      - name: worker
        image: my-worker:v1.2.3
        env:

>         - name: TEMPORAL_HOST_PORT

Reviewer 2: ```suggestion
        - name: TEMPORAL_ADDRESS
```

        # No versioning environment variables
```

**Current Deployment Process:**
1. Build new worker image
2. Update Deployment with new image
3. Kubernetes rolls out new pods, terminating old ones
4. **Risk**: Running workflows may fail if code changes break compatibility

### After: Versioned Workers with Controller

```yaml
# Single Custom Resource manages multiple versions of a worker deployment automatically
apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: my-worker
spec:
  replicas: 3
  workerOptions:
    connection: production-temporal
    temporalNamespace: production
  rollout:
    strategy: Progressive  # Gradual rollout of new versions
    steps:
      - rampPercentage: 10
        pauseDuration: 5m
      - rampPercentage: 50
        pauseDuration: 10m
  sunset:
    scaledownDelay: 1h
    deleteDelay: 24h
  template:
    spec: # Any changes to this spec will trigger the controller to deploy a new version.
      containers:
      - name: worker
        image: my-worker:v1.2.4  # This is the most common value to change, as you roll out a new worker image.
        # Note: Controller automatically adds versioning environment variables:

>         # TEMPORAL_HOST_PORT, TEMPORAL_NAMESPACE, TEMPORAL_DEPLOYMENT_NAME, TEMPORAL_WORKER_BUILD_ID

Reviewer 2: ```suggestion
        # TEMPORAL_ADDRESS, TEMPORAL_NAMESPACE, TEMPORAL_DEPLOYMENT_NAME, TEMPORAL_WORKER_BUILD_ID
```

  oci://docker.io/temporalio/temporal-worker-controller
  
```

### Step 2: Create TemporalConnection Resources

Define connection parameters to your Temporal server(s):

```yaml
apiVersion: temporal.io/v1alpha1
kind: TemporalConnection
metadata:
  name: production-temporal
  namespace: default
spec:
  hostPort: "production.abc123.tmprl.cloud:7233"
  mutualTLSSecret: temporal-cloud-mtls  # If using mTLS
---
apiVersion: temporal.io/v1alpha1
kind: TemporalConnection
metadata:
  name: staging-temporal
  namespace: default
spec:
  hostPort: "staging.abc123.tmprl.cloud:7233"
  mutualTLSSecret: temporal-cloud-mtls
```

### Step 3: Prepare Your Worker Code

Update your worker initialization code to properly handle versioning:

**Before (Unversioned):**
```go
// Worker connects without versioning
worker := worker.New(client, "my-task-queue", worker.Options{})
```

**After (Versioned):**
```go
// Worker must use the build ID/deployment name from environment
// These are set on the deployment by the controller
buildID := os.Getenv("TEMPORAL_WORKER_BUILD_ID")
deploymentName := os.Getenv("TEMPORAL_DEPLOYMENT_NAME")
if buildID == "" || deploymentName == "" {
  // exit with an error
}

> workerOptions := worker.Options{}

Reviewer 2: maybe instead of starting with empty options, we should recommend that users start with the SDK-loaded default options so that the worker has the right connection info
https://github.com/temporalio/temporal-worker-controller/blob/071789e70726ccd88e95c83eaa77536daad6e0a9/internal/demo/util/client.go#L37

# Migrating from Unversioned to Versioned Workflows with Temporal Worker Controller

> This guide helps teams migrate from unversioned Temporal workflows to versioned workflows using the Temporal Worker Controller. It assumes you are currently running workers without Temporal's Worker Versioning feature and want to adopt versioned deployments for safer, more controlled rollouts.

Reviewer 2: ```suggestion
This guide helps teams migrate from unversioned Temporal workflows to versioned workflows using the Temporal Worker Controller. It assumes you are currently running workers without Temporal's Worker Versioning feature and want to adopt versioned worker deployments for safer, more controlled rollouts.
```

Versioned workflows with the Temporal Worker Controller solve these problems by:

- ✅ **Safe Deployments**: New versions run alongside old ones, ensuring running workflows complete successfully
- ✅ **Automated Rollouts**: Progressive rollout strategies reduce risk of new deployments
- ✅ **Easy Rollbacks**: Can instantly route new workflows back to previous versions
- ✅ **Workflow Continuity**: Running workflows can continue on their original version until completion

## Prerequisites

Before starting the migration, ensure you have:

- ✅ **Unversioned Temporal workers**: Currently running workers without Worker Versioning
- ✅ **Kubernetes cluster**: Running Kubernetes 1.19+ with CustomResourceDefinition support
- ✅ **Basic worker configuration**: Workers connect to Temporal with namespace and task queue configuration
- ✅ **Administrative access**: Ability to install Custom Resource Definitions and controllers
- ✅ **Deployment pipeline**: Existing CI/CD system that can be updated to use new deployment method

### Current Environment Variables

Your workers are likely configured with basic environment variables like:

```bash

> TEMPORAL_HOST_PORT=your-namespace.tmprl.cloud:7233

Reviewer 2: ```suggestion
TEMPORAL_HOST_PORT=your-temporal-namespace.tmprl.cloud:7233
```

TEMPORAL_NAMESPACE=your-temporal-namespace
# No TEMPORAL_DEPLOYMENT_NAME or WORKER_BUILD_ID yet
```

The controller will automatically add the versioning-related environment variables during migration.

## Understanding the Differences

### Before: Unversioned Workers

```yaml
# Single deployment, all workers run the same code
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-worker
spec:
  replicas: 3
  template:
    spec:
      containers:
      - name: worker
        image: my-worker:v1.2.3
        env:
        - name: TEMPORAL_HOST_PORT
          value: "production.tmprl.cloud:7233"
        - name: TEMPORAL_NAMESPACE
          value: "production"

>         # No versioning environment variables

Reviewer 2: this example has no mTLS, which is fine, but do we want to demonstrate that at some point (maybe future PR) to show that aspect of the `TemporalConnection` resource?

Author: Added envconfig env vars, I think this is all that's needed?

```

**Current Deployment Process:**
1. Build new worker image
2. Update Deployment with new image
3. Kubernetes rolls out new pods, terminating old ones
4. **Risk**: Running workflows may fail if code changes break compatibility

### After: Versioned Workers with Controller

```yaml

> # Single CRD manages multiple versions automatically

Reviewer 2: ```suggestion
# Single Custom Resource manages multiple versions of a worker deployment automatically
```

apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: my-worker
spec:
  replicas: 3
  workerOptions:
    connection: production-temporal
    temporalNamespace: production
  cutover:
    strategy: Progressive  # Gradual rollout of new versions
    steps:
      - rampPercentage: 10
        pauseDuration: 5m
      - rampPercentage: 50
        pauseDuration: 10m
  sunset:
    scaledownDelay: 1h
    deleteDelay: 24h
  template:
    spec:
      containers:
      - name: worker

>         image: my-worker:v1.2.4  # Update this to deploy new versions

Reviewer 2: maybe this comment or another comment could incorporate some thing like "Updating any part of the pod spec will generate a new Target Build ID for the worker deployment"

Reviewer 2: I just want people to know that too

Author: Reworded, see what you think.

        # Note: Controller automatically adds versioning environment variables:
        # TEMPORAL_HOST_PORT, TEMPORAL_NAMESPACE, TEMPORAL_DEPLOYMENT_NAME, WORKER_BUILD_ID
```

**New Deployment Process:**
1. Build new worker image  

> 2. Update `TemporalWorkerDeployment` CRD with new image

Reviewer 2: ```suggestion
2. Update `TemporalWorkerDeployment` CRD instance with new image
```

3. Controller creates new Kubernetes `Deployment` for the new version

> 4. Controller gradually routes new workflows to new version

Reviewer 2: ```suggestion
4. Controller gradually routes new workflows and existing AutoUpgrade workflows to new version
```

> 5. Old version continues handling existing workflows until they complete

Reviewer 2: ```suggestion
5. Old version continues handling existing Pinned workflows until they complete
```

6. **Safety**: No disruption to running workflows, automated rollout control

**Key Benefits:**
- ✅ **Zero-disruption deployments** - Running workflows continue on original version
- ✅ **Automated version management** - Controller handles registration and routing
- ✅ **Progressive rollouts** - Gradual traffic shifting with automatic pause points
- ✅ **Easy rollbacks** - Instantly route new workflows back to previous version
- ✅ **Workflow continuity** - Deterministic execution preserved across deployments

## Migration Strategy

### Safe Migration Approach

The migration from unversioned to versioned workflows requires careful planning to avoid disrupting running workflows. The key is to transition gradually while maintaining workflow continuity.

**Key Principles:**

> - **Start with Manual strategy** - Prevents automatic promotions during initial setup

Reviewer 2: I think we want folks to try out the thing that they will want to do long-term. So ideally they should try out Progressive, but perhaps with a more conservative rollout policy than they might do later on.

If people start with Manual, they'll have to do "wasted" work developing a system to call the upgrade APIs with the CLI or SDK, which is work that won't be re-used in production, and prone to errors.

Idk that's just my take. I think in an ephemeral environment, folks can happily play with Manual mode all they want, but my recomendation for migrating brownfield environment would be to /add/ a TWD while still deploying the unversioned pollers.

Reviewer 2: just bumping this, I think the "start with manual strategy" recommendation should be removed! but happy to hear your reasoning if you disagree

Author: Adjusted to progressive.

- **Migrate one worker deployment at a time** - Reduces risk and allows learning
- **Test thoroughly in non-production** - Validate the approach before production migration
- **Preserve running workflows** - Ensure in-flight workflows complete successfully
- **Enable automation gradually** - Move to Progressive rollouts only after validation

### Migration Phases

#### Phase 1: Preparation
1. **Install the controller** in non-production environments

> 2. **Update worker code** to support versioning (if needed)

Reviewer 2: the "if needed" feels weird here given that we said that the workers were unversioned as a requirement

3. **Test migration process** with non-critical workers
4. **Prepare CI/CD pipeline changes** for new deployment method

#### Phase 2: Initial Migration
1. **Choose lowest-risk worker** to migrate first

> 2. **Create `TemporalWorkerDeployment` CRD** with Manual strategy

Reviewer 2: ```suggestion
2. **Create `TemporalWorkerDeployment` CRD instance** with Manual strategy
```

3. **Validate controller management** works correctly
4. **Update deployment pipeline** for this worker

#### Phase 3: Gradual Rollout
1. **Migrate remaining workers** one at a time
2. **Enable Progressive rollouts** for validated workers
3. **Monitor and tune** rollout configurations
4. **Train team** on new deployment process

> ### Recommended Migration Order

Reviewer 2: i like this! I think its helpful for users to know they don't have to swap over all at once

1. **Background/batch processing workers** - Lower risk, easier to validate
2. **Internal service workers** - Limited external impact
3. **Customer-facing workers** - Highest risk, migrate last with most care

## Step-by-Step Migration

### Step 1: Install the Temporal Worker Controller

```bash
# Install the controller using Helm
helm install -n temporal-system --create-namespace \
  temporal-worker-controller \
  oci://docker.io/temporalio/temporal-worker-controller
  
```

### Step 2: Create TemporalConnection Resources

Define connection parameters to your Temporal server(s):

```yaml
apiVersion: temporal.io/v1alpha1
kind: TemporalConnection
metadata:
  name: production-temporal
  namespace: default
spec:
  hostPort: "production.abc123.tmprl.cloud:7233"
  mutualTLSSecret: temporal-cloud-mtls  # If using mTLS
---
apiVersion: temporal.io/v1alpha1
kind: TemporalConnection
metadata:
  name: staging-temporal
  namespace: default
spec:
  hostPort: "staging.abc123.tmprl.cloud:7233"
  mutualTLSSecret: temporal-cloud-mtls
```

> ### Step 3: Prepare Your Worker Code (If Needed)

Reviewer 2: same re: "if needed"

> Most workers will work without changes, but you may need to update your worker initialization code to properly handle versioning:

Reviewer 2: if they are unversioned, they won't work without changes

**Before (Unversioned):**
```go
// Worker connects without versioning
worker := worker.New(client, "my-task-queue", worker.Options{})
```

**After (Versioned - Optional Enhancement):**
```go

> // Worker can optionally use build ID from environment

Reviewer 2: worker MUST use build id and deployment name from environment

buildID := os.Getenv("WORKER_BUILD_ID")
workerOptions := worker.Options{}
if buildID != "" {
    workerOptions.BuildID = buildID

>     workerOptions.UseBuildIDForVersioning = true

Reviewer 2: I think this is wrong too and it's actually worker.Options.DeploymentOptions{BuildId, DeploymentName}, but double check the docs or the demo worker code

}
worker := worker.New(client, "my-task-queue", workerOptions)
```

> > **Note**: The controller automatically sets `WORKER_BUILD_ID` environment variable, so most workers will work without code changes.

Reviewer 2: maybe this is the write place to say that WORKER_BUILD_ID and WORKER_DEPLOYMENT_NAME (double check those) are generated by the controller and must be passed to the worker's worker.Options.DeploymentOptions

### Step 4: Create Your First TemporalWorkerDeployment

> Start with your lowest-risk worker. Convert your existing unversioned Deployment to a `TemporalWorkerDeployment` CRD:

Reviewer 2: ```suggestion
Start with your lowest-risk worker. Convert your existing unversioned Deployment to a `TemporalWorkerDeployment` CRD instance:
```
I think we want to tell people to keep their unversioned workers around, so maybe "Make a copy of your existing unversioned Deployment and convert it to a ..." would be extra clear that we don't recommend to replace them?

> **Current Unversioned Deployment:**

Reviewer 2: ```suggestion
**Existing Unversioned Deployment:**
```
just because the term "Current" is so overloaded in the versioning world? wdyt?

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: payment-processor
spec:
  replicas: 3
  template:
    spec:
      containers:
      - name: worker
        image: payment-processor:v1.5.2
        env:
        - name: TEMPORAL_HOST_PORT
          value: "production.tmprl.cloud:7233"
        - name: TEMPORAL_NAMESPACE
          value: "production"
```

> **New TemporalWorkerDeployment CRD (IMPORTANT: Use Manual strategy initially):**

Reviewer 2: ```suggestion
**New TemporalWorkerDeployment resource:**
```
or "instance"?

```yaml
apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: payment-processor
  labels:
    app: payment-processor
spec:
  replicas: 3
  workerOptions:
    connection: production-temporal
    temporalNamespace: production

>   # CRITICAL: Use Manual during initial migration

Reviewer 2: idk about this, lets discuss monday?

>   cutover:

Reviewer 2: ```suggestion
  rollout:
```

    strategy: Manual
  sunset:
    scaledownDelay: 30m
    deleteDelay: 2h
  template:
    spec:
      containers:
      - name: worker

>         image: payment-processor:v1.5.2  # Same image as current deployment

Reviewer 2: ```suggestion
        image: payment-processor:v1.5.2  # Same image as current deployment to ensure no breaking changes
```

        resources:
          requests:
            memory: "512Mi"
            cpu: "250m"
        # Note: Controller automatically adds TEMPORAL_* env vars
```

### Step 5: Deploy the TemporalWorkerDeployment

> 1. **Create the `TemporalWorkerDeployment` CRD:**

Reviewer 2: ```suggestion
1. **Create the `TemporalWorkerDeployment` custom resource:**
```
or CRD instance, or resource

   ```bash
   kubectl apply -f payment-processor-versioned.yaml
   ```

2. **Wait for the version to be registered:**
   ```bash

>    # Monitor until status shows the version is registered and current

Reviewer 2: just a note that if mode is manual, version won't become current. but if mode is progressive (which imo it should be) then it will.

4. **Verify in Temporal UI:**
   - Check the Workers page in Temporal UI
   - You should see your worker deployment with version information
   - New workflows should be routed to the versioned worker

### Step 6: Transition from Unversioned to Versioned

Now you need to carefully transition from your old unversioned deployment to the new versioned one:

1. **Ensure both deployments are running:**
   ```bash
   # Check your original unversioned deployment
   kubectl get deployment payment-processor
   
   # Check the new controller-managed versioned deployment  
   kubectl get deployments -l temporal.io/managed-by=temporal-worker-controller
   ```

2. **Monitor workflow routing:**
   - In Temporal UI, check that new workflows are being routed to the versioned worker
   - Existing workflows should continue on the unversioned worker until they complete

3. **Wait for existing workflows to complete:**
   ```bash
   # Monitor running workflows in Temporal UI
   # Or use Temporal CLI to check workflow status
   temporal workflow list --namespace production
   ```

4. **Scale down the original unversioned deployment:**
   ```bash
   # Gradually reduce replicas of the original deployment

>    kubectl scale deployment payment-processor --replicas=1

Reviewer 2: what's the point of scaling to 1?

   
   # Monitor that workflows continue to work correctly
   # Once confident, scale to zero
   kubectl scale deployment payment-processor --replicas=0
   ```

5. **Clean up the original deployment:**
   ```bash
   # Only after confirming all workflows are handled by versioned workers
   kubectl delete deployment payment-processor
   ```

### Step 7: Enable Automated Rollouts

Once the initial migration is complete and validated, enable automated rollouts for future deployments:

```bash

> # Update the TemporalWorkerDeployment CRD to use Progressive strategy

Reviewer 2: ```suggestion
# Update the TemporalWorkerDeployment custom resource to use Progressive strategy
```

kubectl patch temporalworkerdeployment payment-processor --type='merge' -p='{
  "spec": {
    "cutover": {
      "strategy": "Progressive",
      "steps": [
        {"rampPercentage": 10, "pauseDuration": "5m"},
        {"rampPercentage": 50, "pauseDuration": "10m"}
      ]
    }
  }
}'
```

### Step 8: Update Your CI/CD Pipeline

Modify your deployment pipeline to work with the new versioned approach:

**Before (Unversioned):**
```bash
# Old pipeline updated Deployment directly
kubectl set image deployment/payment-processor worker=payment-processor:v1.6.0
```

**After (Versioned):**
```bash

> # New pipeline updates TemporalWorkerDeployment CRD

Reviewer 2: ```suggestion
# New pipeline updates TemporalWorkerDeployment custom resource
```

4. Gradually routes traffic according to Progressive strategy
5. Scales down and cleans up old version once new version is fully deployed

### Step 9: Test Your First Versioned Deployment

Deploy a new version to validate the entire flow:

1. **Make a small, safe change** to your worker code
2. **Build and push** a new container image
3. **Update the TemporalWorkerDeployment:**
   ```bash
   kubectl patch temporalworkerdeployment payment-processor --type='merge' -p='{"spec":{"template":{"spec":{"containers":[{"name":"worker","image":"payment-processor:v1.6.0"}]}}}}'
   ```
4. **Monitor the rollout:**
   ```bash
   # Watch the deployment progress
   kubectl get temporalworkerdeployment payment-processor -w
   
   # Check that new deployment is created
   kubectl get deployments -l temporal.io/managed-by=temporal-worker-controller
   ```
5. **Verify in Temporal UI** that traffic is gradually shifting to the new version

## Configuration Reference

### Rollout Strategies

See the [Concepts](concepts.md) document for detailed explanations of rollout strategies. Here are the basic configuration patterns:

**Manual Strategy (Default Behavior):**
```yaml

> cutover:

Reviewer 2: replace cutover with rollout for all of these

  strategy: Manual
# Requires manual intervention to promote versions
```

**Immediate Cutover:**
```yaml
cutover:
  strategy: AllAtOnce
# Immediately routes 100% traffic to new version when healthy
```

**Progressive Rollout:**
```yaml
cutover:
  strategy: Progressive
  steps:
    - rampPercentage: 1
      pauseDuration: 5m
    - rampPercentage: 10  
      pauseDuration: 10m
    - rampPercentage: 50
      pauseDuration: 15m
  gate:
    workflowType: "HealthCheck"  # Optional validation workflow
```

### Sunset Configuration

```yaml
sunset:
  scaledownDelay: 1h    # Wait 1 hour after draining before scaling to 0
  deleteDelay: 24h      # Wait 24 hours after draining before deleting
```

### Resource Management

**CPU and Memory:**
```yaml
template:
  spec:
    containers:
    - name: worker
      resources:
        requests:
          memory: "1Gi"
          cpu: "500m"
        limits:
          memory: "2Gi" 
          cpu: "1"
```

**Pod Annotations and Labels:**
```yaml
template:
  metadata:
    labels:
      team: payments
      environment: production

> ```

Reviewer 2: I feel like this is out of scope, people should read k8s docs on pod spec for this

## Testing and Validation

### Pre-Migration Testing

1. **Test in non-production environment:**
   ```bash
   # Create test TemporalWorkerDeployment
   kubectl apply -f test-worker.yaml
   
   # Verify worker registration
   kubectl logs -l temporal.io/worker-deployment=test-worker
   ```

2. **Validate environment variables:**
   ```bash
   kubectl exec -it deployment/test-worker-v1 -- env | grep TEMPORAL
   ```

### Post-Migration Validation

1. **Check version status:**
   ```bash
   kubectl get temporalworkerdeployment -o wide
   ```

2. **Monitor version transitions:**
   ```bash
   kubectl get events --field-selector involvedObject.kind=TemporalWorkerDeployment
   ```

3. **Validate workflow routing:**
   - Start test workflows
   - Verify they're routed to correct versions
   - Check Temporal UI for version distribution

## Common Migration Patterns

Here are common scenarios when migrating from unversioned to versioned workflows:

### Pattern 1: Microservices Architecture

> If you have multiple services with their own workers, migrate each service independently:

Reviewer 2: maybe something like "and each service is versioned and patched separately"?

```yaml
# Payment service worker
apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: payment-processor
  namespace: payments
spec:
  workerOptions:
    connection: production-temporal
    temporalNamespace: payments
  cutover:

>     strategy: Progressive  # Conservative rollout for financial operations

Reviewer 2: ```suggestion
    strategy: Progressive
```
we rec Progressive regardless, so no need to mention "conservative" imo

    steps:
      - rampPercentage: 5
        pauseDuration: 10m
      - rampPercentage: 25
        pauseDuration: 15m
  # ... rest of config
---
# Notification service worker (separate service, different risk profile)
apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: notification-sender
  namespace: notifications
spec:
  workerOptions:
    connection: production-temporal
    temporalNamespace: notifications
  cutover:

>     strategy: AllAtOnce  # Lower risk, faster rollouts acceptable

Reviewer 2: maybe "desired" instead of acceptable? IMO if people don't know what they want they should do progressive, so "acceptable" is not a good enough reason to do AllAtOnce

  # ... rest of config
```

### Pattern 2: Environment-Specific Strategies

Use different rollout strategies based on environment risk:

```yaml
# Production - Conservative rollout
apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: order-processor
  namespace: production
spec:
  workerOptions:
    connection: production-temporal
    temporalNamespace: production
  cutover:
    strategy: Progressive
    steps:
      - rampPercentage: 10
        pauseDuration: 15m
      - rampPercentage: 50
        pauseDuration: 30m
    gate:
      workflowType: "HealthCheck"  # Validate new version before proceeding
---
# Staging - Fast rollout for testing
apiVersion: temporal.io/v1alpha1
kind: TemporalWorkerDeployment
metadata:
  name: order-processor
  namespace: staging
spec:
  workerOptions:
    connection: staging-temporal
    temporalNamespace: staging
  cutover:

>     strategy: AllAtOnce  # Immediate cutover for faster iteration

Reviewer 2: maybe just "faster rollout"? I feel like iteration speed is dependent on so many factors, least of which is how long it takes to deploy one service to one env

```

### Pattern 3: Gradual Team Migration

Migrate teams/services based on their readiness and risk tolerance:

**Phase 1: Low-Risk Services**
- Background processing workers
- Internal tooling workflows
- Non-customer-facing operations

**Phase 2: Medium-Risk Services**  
- Internal API workflows
- Data processing pipelines
- Administrative workflows

**Phase 3: High-Risk Services**
- Customer-facing workflows
- Payment processing
- Critical business operations

## Troubleshooting

### Common Issues

**1. Workers Not Registering with Temporal**

*Symptoms:*
```
status:
  targetVersion:
    status: NotRegistered
```

*Solutions:*
- Check worker logs for connection errors
- Verify TemporalConnection configuration  
- Ensure TLS secrets are properly configured
- Verify network connectivity to Temporal server

> - Check that worker code properly handles `WORKER_BUILD_ID` environment variable

Reviewer 2: and WORKER_DEPLOYMENT_NAME

**2. New Workflows Still Going to Unversioned Workers**

*Symptoms:*
- Temporal UI shows workflows executing on unversioned workers

> - Versioned workers appear idle

Reviewer 2: not sure what "appear idle" would look like, but the Temporal UI shows workflows *not* executing on versioned workers (you can see this by clicking "workflows" from the version line item in the UI)

*Solutions:*
- Verify versioned workers are properly registered in Temporal UI

> - Check that new workflow starts are using the correct task queue

Reviewer 2: ```suggestion
- Check that new workflows are starting on the correct task queue
```
